zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -2119,8 +2126,41 @@ private String getHexSessionId() {
    * 2. create mode is EPHEMERAL or EPHEMERAL_SEQUENTIAL
    */
   private boolean isSessionAwareOperation(String expectedSessionId, CreateMode mode) {
-    return expectedSessionId != null && !expectedSessionId.isEmpty() && (
-        mode == CreateMode.EPHEMERAL || mode == CreateMode.EPHEMERAL_SEQUENTIAL);
+    return expectedSessionId != null && !expectedSessionId.isEmpty() && mode.isEphemeral();
+  }
+
+  private ZooKeeper getExpectedZookeeper(final String expectedSessionId) {
```

Comment:
This is a great idea!

Reply:
I actually also tried to use this API in `create()` that we previous added. But inside `create()` there is a `isSessionAwareOperation()` check. If we don't need this check, we could just use the this method `getExpectedZookeeper`. And I actually don't think we need the check "isSessionAwareOperation()" since we only expose `createEphemeral()` to have expectedSession. We should be good.

Reply:
How do we reason this is correct? Note, previously, we have `acquireEventLock` when checking sessionId is equal to expected sessionId.
```
if (isSessionAwareOperation(expectedSessionId, mode)) {
          acquireEventLock();
          try {
            final String actualSessionId = Long.toHexString(zooKeeper.getSessionId());
            if (!actualSessionId.equals(expectedSessionId)) {
              throw new ZkSessionMismatchedException(
                  "Failed to create ephemeral node! There is a session id mismatch. Expected: "
                      + expectedSessionId + ". Actual: " + actualSessionId);
            }

            /*
             * Cache the zookeeper reference and make sure later zooKeeper.create() is being run
             * under this zookeeper connection. This is to avoid locking zooKeeper.create() which
             * may cause potential performance issue.
             */
            zooKeeper = ((ZkConnection) getConnection()).getZookeeper();
```

Reply:
So here, we are saying the native Zookeeper object `zookeeper` can't be changed even session changed. Still one Zookeeeper object one session invariant, right? Save us one lock.

Reply:
@kaisun2000 Yes, correct! We don't have to lock it actually.